src/DynamoCore/DSEngine/NodeToCode.cs

```diff
@@ -1,16 +1,692 @@
+            /// </summary>
+            public string NumberedVariable
+            {
+                get { return Variable + ((ID == 0) ? String.Empty : ID.ToString()); }
+            }
+
+            /// <summary>
+            /// Number suffix.
+            /// </summary>
+            public int ID { get; set; }
+
+            /// <summary>
+            /// Indicate it is for ASTs in a new code block
+            /// </summary>
+            public bool IsNewSession { get; set; }
+
+            /// <summary>
+            /// Constructor.
+            /// </summary>
+            /// <param name="variable"></param>
+            public NumberingState(string variable)
+            {
+                Variable = variable;
+                ID = 0;
+                IsNewSession = false;
+            }
+
+            /// <summary>
+            /// Increase number. 
+            /// </summary>
+            public void BumpID()
+            {
+                ID += 1;
+            }
+        }
+
+        /// <summary>
+        /// Generate short variable name.
+        /// </summary>
+        private class ShortNameGenerator
+        {
+            private NumberingState ns = new NumberingState(AstBuilder.StringConstants.ShortVarPrefix);
+
+            public ShortNameGenerator()
+            {
+
+            }
+            public string GetNextName()
+            {
+                ns.BumpID();
+                return ns.NumberedVariable;
+            }
+        }
+
+        private static void MarkConnectivityForNode(
+            HashSet<NodeModel> selection,
+            bool[,] connectivityMatrx,
+            Dictionary<NodeModel, int> nodeMap,
+            List<NodeModel> path)
+        {
+            var leave = path.Last();
+            var children = leave.OutputNodes.Values.SelectMany(s => s).Select(t => t.Item2);
+            if (children.Any())
+            {
+                foreach (var child in children)
+                {
+                    if (path.Contains(child))
+                        continue;
+
+                    path.Add(child);
+                    MarkConnectivityForNode(selection, connectivityMatrx, nodeMap, path);
+                }
+            }
+            else
+            {
+                var nodes = path.ToList();
+                var idx1 = nodeMap[path.First()];
+
+                int k = nodes.FindIndex(n => !n.IsConvertible || !selection.Contains(n));
+                if (k >= 0)
+                {
+                    for (int i = k + 1; i < nodes.Count; i++)
+                    {
+                        int idx2;
+                        if (!nodeMap.TryGetValue(nodes[i], out idx2))
+                            continue;
+
+                        connectivityMatrx[idx1, idx2] = false;
+                        connectivityMatrx[idx2, idx1] = false;
+                    }
+                }
+            }
+
+            path.RemoveAt(path.Count - 1);
+        }
```

Comment:
sorry that this method doesn't contain the comment, but i put a comment below about what this function does:
```
            // For two convertible nodes if there is path from one to the other
            // goes through some node that is unconvertible or not in the 
            // selection set, these two nodes cannot be in the same coee block,
            // otherwise there will be a circular reference. For example:
            //
            //     a -> x -> b
            //
            // And the final graph would be
            //
            //    a, b <--> x
            // 
            // So here for a node which is convertible and from selection set, 
            // find all its upstream paths, and if any path contains node that 
            // is not in the selection set or is not convertible, mark all 
            // nodes after this node along path as not reachable. 
 ```